### PR TITLE
First draft for TDNA

### PR DESCRIPTION
**Pull request by thesiddharth**

### CODE REVIEW

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.linkedin.norbert.cluster.Node
+import com.linkedin.norbert.logging.Logging
+import com.linkedin.norbert.network.common.Endpoint
+import com.linkedin.norbert.network.partitioned.loadbalancer.{DefaultClusteredLoadBalancer, DefaultClusteredLoadBalancerFactory, PartitionedLoadBalancer}
+import com.linkedin.norbert.norbertutils.SystemClockComponent
+
+/**
+ * This class is intended for applications where there is a mapping from partitions -> servers able to respond to those requests. Requests are round-robined
+ * between the partitions
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala]: This needs better documentation. We should explain what the purpose of such a load balancer is, and what the various parameters like `cycleTime` and `slotTime` mean.

File: cluster/src/main/scala/com/linkedin/norbert/norbertutils/Clock.scala [hunk 2]
```diff
@@ -41,3 +41,7 @@ object SystemClock extends Clock {
 object SystemClockComponent extends ClockComponent {
```
Comment (karanparikh) [inline cluster/src/main/scala/com/linkedin/norbert/norbertutils/Clock.scala]: (not your code) I think this can be removed since it is not being used. All we need is the `trait`

File: network/src/main/scala/com/linkedin/norbert/network/client/loadbalancer/LoadBalancerFactory.scala [hunk 3]
```diff
@@ -18,6 +18,8 @@ package network
 package client
 package loadbalancer
 
+import java.util.concurrent.atomic.AtomicBoolean
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/client/loadbalancer/LoadBalancerFactory.scala]: This doesn't seem to be used.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClientFactory.scala [hunk 4]
```diff
@@ -48,7 +48,7 @@ class PartitionedNetworkClientFactory[PartitionedId](clientName: String,
       if(retryStrategy == null) 
         throw new IllegalArgumentException("Retry strategy needs to be provided if you enable selective retry")
       else
-	config.retryStrategy = Some(retryStrategy)
+	      config.retryStrategy = Some(retryStrategy)
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClientFactory.scala]: Revert this change.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultClusteredLoadBalancerFactory.scala [hunk 5]
```diff
@@ -40,264 +38,163 @@ abstract class DefaultClusteredLoadBalancerFactory[PartitionedId](numPartitions:
...
-     */
-    protected def generateClusterToNodeMap(nodes: Set[Endpoint]): Map[Int, IndexedSeq[Endpoint]] = {
-      val clusterToNodeMap = (for (n <- nodes) yield (n, clusterId(n.node)))
-        .foldLeft(Map.empty[Int, IndexedSeq[Endpoint]]) {
-        case (map, (node, key)) => map + (key -> (node +: map.get(key).getOrElse(Vector.empty[Endpoint])))
-      }
-
-      clusterToNodeMap
-    }
+abstract class DefaultClusteredLoadBalancer[PartitionedId](endpoints: Set[Endpoint], partitionForId: PartitionedId => Int, clusterId: Node => Int, numPartitions: Int, serveRequestsIfPartitionMissing: Boolean)
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultClusteredLoadBalancerFactory.scala]: Documentation

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultLoadBalancerHelper.scala [hunk 6]
```diff
@@ -101,16 +94,12 @@ trait DefaultLoadBalancerHelper extends LoadBalancerHelpers with Logging {
...
     }
   }
 
-  def compensateCounter(idx: Int, count:Int, counter:AtomicInteger) {
-    if (idx + 1 + count <= 0) {
-      // Integer overflow
-      counter.set(idx + 1 - java.lang.Integer.MAX_VALUE + count)
-    }
-    counter.set(idx + 1 + count)
+  def isEndpointViable(capability: Option[Long], persistentCapability: Option[Long], endpoint: Endpoint): Boolean = {
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultLoadBalancerHelper.scala]: This can be a `val` rather than a function.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2009-2010 LinkedIn, Inc
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala]: Update the year.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareLoadBalancerHelper.scala [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            , n) :: offsetNodeTuples
+
+    offsetNodeTuples.foldLeft(Map.empty[Int, Set[Endpoint]]) {
+      case (map, (offset, node)) => map + (offset -> (map.getOrElse(offset, Set.empty[Endpoint]) + node))
+    }
+  }
+
+  def validateOffsets(nodes: Set[Endpoint]): Unit = {
+
+    for(n <- nodes) if (n.node.offset.isEmpty) throw new InvalidClusterException("Node %d doesn't have a GC offset".format(n.node.id))
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareLoadBalancerHelper.scala]: `foreach` instead?

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwarePartitionedLoadBalancerFactory.scala [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * This class is intended for applications where there is a mapping from partitions -> servers able to respond to those requests. Requests are round-robined
+ * between the partitions
+ */
+abstract class GcAwarePartitionedLoadBalancerFactory[PartitionedId](numPartitions: Int, cycleTime: Int, slotTime: Int, serveRequestsIfPartitionMissing: Boolean = true)
+        extends DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartitions, serveRequestsIfPartitionMissing) with Logging {
+
+  override def newLoadBalancer(endpoints: Set[Endpoint]): PartitionedLoadBalancer[PartitionedId] = GcAwarePartitionedLoadBalancer(endpoints, partitionForId, numPartitions, cycleTime, slotTime, serveRequestsIfPartitionMissing)
+
+//  class TestEndpoint(val node: Node, var csr: Boolean) extends Endpoint {
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwarePartitionedLoadBalancerFactory.scala]: Please remove.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwarePartitionedLoadBalancerFactory.scala [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      val gcCycleTime = cycleTime
+      val gcSlotTime = slotTime
+    }
+  }
+}
+
+abstract class GcAwarePartitionedLoadBalancer[PartitionedId](endpoints: Set[Endpoint], partitionForId: PartitionedId => Int, numPartitions: Int, cycleTime: Int, slotTime: Int, serveRequestsIfPartitionMissing: Boolean)
+        extends DefaultPartitionedLoadBalancer[PartitionedId](endpoints, partitionForId, numPartitions, serveRequestsIfPartitionMissing) with GcAwareLoadBalancerHelper with SystemClockComponent {
+
+//  override def generatePartitionToNodeMap(nodes: Set[Endpoint], numPartitions: Int, serveRequestsIfPartitionMissing: Boolean) = super[GcAwareLoadBalancerHelper].generatePartitionToNodeMap(nodes, numPartitions, serveRequestsIfPartitionMissing)
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwarePartitionedLoadBalancerFactory.scala]: Remove

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import com.linkedin.norbert.logging.Logging
+import com.linkedin.norbert.network.common.Endpoint
+import com.linkedin.norbert.network.partitioned.loadbalancer.{DefaultClusteredLoadBalancer, DefaultClusteredLoadBalancerFactory, PartitionedLoadBalancer}
+import com.linkedin.norbert.norbertutils.SystemClockComponent
+
+/**
+ * Extends DefaultClusteredLoadBalancer to add GC-awareness.
+ * Nodes are additionally filtered to exclude those that are currently garbage collecting.
+ *
+ * @param cycleTime: The time period in which each node in the colo undergoes garbage collection exactly once
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala]: I think "cluster" rather than "colo" is a more appropriate term here.

Comment (thesiddharth) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala]: Every machine in the data center will GC once in 'cycleTime', each cluster will GC in one 'slot' of duration 'slotTime'.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+ * @param slotTime: The time allotted for the nodes in one cluster to finish pending requests and garbage collect.
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala]: "Time required to serve a request + time required to garbage collect"
We should also specify time units for both `cycleTime` and `slotTime`.

File: cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala [hunk 13]
```diff
@@ -77,8 +81,9 @@ object Node {
  * @param available whether or not the node is currently able to process requests
  * @param partitions the partitions for which the node can handle requests
  * @param capability the 64 bits Long representing up to 64 node capabilities
+ * @param offset the offset in the time cycle at which this node GCs
```
Comment (karanparikh) [inline cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala]: This needs better documentation. An example might be useful here. We should also state that this is only used if we are doing GC aware load balancing.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala (hunk 11 above)
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/gcaware/GcAwareClusteredLoadBalancerFactory.scala]: In that case use let's use "data center" instead of "colo".

File: network/src/main/scala/com/linkedin/norbert/network/netty/GC.scala [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package com.linkedin.norbert.network.netty
+
+/**
+ * Created by sishah on 7/2/15.
+ */
+class GC extends Runnable{
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/netty/GC.scala]: Please come up with a better name than `GC`

File: network/src/main/scala/com/linkedin/norbert/network/netty/GcParamWrapper.scala [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Created by sishah on 7/6/15.
+ *
+ * A wrapper object over three GC parameters -
+ *  - slaTime: The SLA request time.
+ *  - cycleTime: The time period in which each node in the data center undergoes garbage collection exactly once
+ *  - slotTime: The time allotted for the nodes in one cluster to finish pending requests and garbage collect
+ *
+ *  GC Awareness on the server side kicks in only if all three have non-zero values, and the node bound to the server
+ *  has a non-zero offset.
```
Comment (karanparikh) [inline network/src/main/scala/com/linkedin/norbert/network/netty/GcParamWrapper.scala]: Non-zero offset is not being checked here though.

File: cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala [hunk 16]
```diff
@@ -42,9 +42,9 @@ object Node {
       val partitions = node.getPartitionList.asInstanceOf[java.util.List[Int]].foldLeft(Set[Int]()) { (set, i) => set + i }
 
       if(!node.hasPersistentCapability)
```
Comment (calcu16) [inline cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala]: With this many if else cases, it might be worthwhile to individually calculate the arguments e.g.
val persistentCapability = if (node.hasPersistenCapability) Some(...) else None
val offset = if (...) Some(...) else None

File: cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala [hunk 17]
```diff
@@ -89,7 +95,7 @@ final case class Node(id: Int, url: String, available: Boolean, partitionIds: Se
     case _ => false
   }
 
-  override def toString = "Node(%d,%s,[%s],%b,0x%08X,0x%08X)".format(id, url, partitionIds.mkString(","), available, if (capability.isEmpty) 0L else capability.get, if (persistentCapability.isEmpty) 0L else persistentCapability.get)
+  override def toString = "Node(%d,%s,[%s],%b,0x%08X,0x%08X,%d)".format(id, url, partitionIds.mkString(","), available, if (capability.isEmpty) 0L else capability.get, if (persistentCapability.isEmpty) 0L else persistentCapability.get, if (offset.isEmpty) 0 else offset.get)
```
Comment (calcu16) [inline cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala]: I think you can have a 0 offset, maybe a different sentinal value would be better (e.g. -1)

File: cluster/src/main/scala/com/linkedin/norbert/norbertutils/Clock.scala [hunk 18]
```diff
@@ -38,6 +38,6 @@ object SystemClock extends Clock {
   def getCurrentTimeMilliseconds = System.currentTimeMillis
 }
 
-object SystemClockComponent extends ClockComponent {
```
Comment (calcu16) [inline cluster/src/main/scala/com/linkedin/norbert/norbertutils/Clock.scala]: This doesn't seem like a backwards compatible change

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala [hunk 19]
```diff
@@ -25,7 +25,6 @@ import netty.NettyPartitionedNetworkClient
 import client.NetworkClientConfig
 import cluster.{Node, ClusterDisconnectedException, InvalidClusterException, ClusterClientComponent}
 import scala.util.Random
-import java.util
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala]: While I appreciate the cleanup it would be better for unrelated changes to be in a separate pull request

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultLoadBalancerHelper.scala (hunk 6 above)
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultLoadBalancerHelper.scala]: def's can be overriden with dynamic results, this way might actually be better (and a val is going to be a getter function in java anyway)

File: network/src/main/scala/com/linkedin/norbert/network/server/MessageExecutorComponent.scala [hunk 20]
```diff
@@ -99,6 +155,16 @@ class ThreadPoolMessageExecutor(clientName: Option[String],
   def executeMessage[RequestMsg, ResponseMsg](request: RequestMsg, responseHandler:  Option[(Either[Exception, ResponseMsg]) => Unit], context: Option[RequestContext] = None)
                                              (implicit is: InputSerializer[RequestMsg, ResponseMsg]) {
     val rr = new RequestRunner(request, requestTimeout, context, filters, responseHandler, is = is)
+
+    //Reject messages that arrive post the GC start period.
+    if(enableGcAwareness && isCurrentlyDownToGC(myNode.get.offset.get)) {
+      statsActor.endRequest(0, rr.id)
+
+      totalNumRejected.incrementAndGet
+      log.warn("Rejecting request in favour of going down to GC. Queue size is currently " + requestQueue.size)
```
Comment (calcu16) [inline network/src/main/scala/com/linkedin/norbert/network/server/MessageExecutorComponent.scala]: I'm not sure I care about queue size here

File: cluster/src/main/scala/com/linkedin/norbert/norbertutils/Clock.scala (hunk 18 above)
Comment (thesiddharth) [inline cluster/src/main/scala/com/linkedin/norbert/norbertutils/Clock.scala]: The object wasn't used anywhere.

File: cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala (hunk 17 above)
Comment (thesiddharth) [inline cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala]: Yup, I'll replace 0 here with -1. I'm checking if the offset parameter is set during protobuf decoding, so this isn't an issue there.

File: network/src/main/scala/com/linkedin/norbert/network/server/MessageExecutorComponent.scala (hunk 20 above)
Comment (thesiddharth) [inline network/src/main/scala/com/linkedin/norbert/network/server/MessageExecutorComponent.scala]: Yup, will remove.

File: cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala (hunk 16 above)
Comment (thesiddharth) [inline cluster/src/main/scala/com/linkedin/norbert/cluster/Node.scala]: Fixed

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala (hunk 19 above)
Comment (thesiddharth) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/PartitionedNetworkClient.scala]: Done. Kinda. As discussed.

**Comment by calcu16** [comment]
Ship It.
